types/key/util.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package key
+
+import (
+	crand "crypto/rand"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+
+	"go4.org/mem"
+)
+
+// rand fills b with cryptographically strong random bytes. Panics if
+// no random bytes are available.
+func rand(b []byte) {
+	if _, err := io.ReadFull(crand.Reader, b[:]); err != nil {
+		panic("unable to read random bytes from OS")
+	}
+}
+
+// clamp25519 clamps b, which must be a 32-byte Curve25519 private
+// key, to a safe value.
+//
+// The clamping effectively constrains the key to a number between
+// 2^251 and 2^252-1, which is then multiplied by 8 (the cofactor of
+// Curve25519). This produces a value that doesn't have any unsafe
+// properties when doing operations like ScalarMult.
+//
+// See
+// https://web.archive.org/web/20210228105330/https://neilmadden.blog/2020/05/28/whats-the-curve25519-clamping-all-about/
+// for a more in-depth explanation of the constraints that led to this
+// clamping requirement.
+//
+// PLEASE NOTE that not all Curve25519 curve values require
+// clamping. When implementing a new key type that uses Curve25519,
+// you must evaluate whether that particular key's use requires
+// clamping. Here are some existing uses and whether you should clamp
+// private keys at creation.
+//
+//  - NaCl box: yes, clamp at creation.
+//  - WireGuard: yes, clamp at creation.
```

Comment:
I'm not sure this is true. I believe wireguard will accept any key and clamp it before use. The idea is any random bytes can be used as a wireguard key.

Reply:
Hm, you're right. The thing that confused me was that we imported the clamping from wireguard-go during some refactor, because at one point we were talking to wireguard-go with our own Go structs, and so had to safely initialize our own keys.

However, now, wireguard-go only ever ingests private keys through uapi, which clamps private keys during the decoding from hex.

Updated the docstring, and will adjust our key creation when I get to that key type.